src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -1,60 +1,1057 @@
+        ctx.Document.Definitions
+        |> List.fold (fun acc def ->
+            let path = match def.Name with | Some name -> [name] | None -> []
+            match def with
+            | OperationDefinition odef -> checkDirectivesInOperation acc ctx.Schema fragmentDefinitions path visitedFragments odef
+            | FragmentDefinition frag -> checkDirectivesInFragment acc ctx.Schema fragmentDefinitions path visitedFragments frag) Success
+
+    let rec private getDirectiveNamesInSelection (path : Path) (selection : Selection) : (Path * string list) list =
+        match selection with
+        | Field field ->
+            let path = field.AliasOrName :: path
+            let fieldDirectives = [ path, field.Directives |> List.map (fun x -> x.Name) ]
+            let selectionSetDirectives = field.SelectionSet |> List.collect (getDirectiveNamesInSelection path)
+            fieldDirectives |> List.append selectionSetDirectives
+        | InlineFragment frag -> getDirectiveNamesInDefinition path (FragmentDefinition frag)
+        | FragmentSpread spread -> [ path, spread.Directives |> List.map (fun x -> x.Name) ]
+
+    and private getDirectiveNamesInDefinition (path : Path) (frag : Definition) : (Path * string list) list =
+        let fragDirectives = [ path, frag.Directives |> List.map (fun x -> x.Name) ]
+        let selectionSetDirectives = frag.SelectionSet |> List.collect (getDirectiveNamesInSelection path)
+        fragDirectives |> List.append selectionSetDirectives
+
+    let internal validateUniqueDirectivesPerLocation (ctx : ValidationContext) =
+        ctx.Definitions
+        |> List.collect (fun def ->
+            let path = match def.Name with | Some name -> [name] | None -> []
+            let defDirectives = path, def.Directives |> List.map (fun x -> x.Name)
+            let selectionSetDirectives = def.Definition.SelectionSet |> List.collect (getDirectiveNamesInSelection path)
+            defDirectives :: selectionSetDirectives)
+        |> List.map (fun (path, directives) -> path, directives |> List.groupBy id |> List.map (fun (k, v) -> k, v.Length))
+        |> List.fold (fun acc (path, directives) ->
+            directives 
+            |> List.filter (fun (_, length) -> length > 1)
+            |> List.fold (fun acc (name, length) -> acc @ Error.AsResult(sprintf "Directive '%s' appears %i times in the location it is used. Directives must be unique in their locations." name length, path)) acc) Success
+
+    let internal validateVariableUniqueness (ctx : ValidationContext) =
+        ctx.Document.Definitions
+        |> List.choose (function OperationDefinition def -> Some (def.Name, def.VariableDefinitions) | _ -> None)
+        |> List.fold (fun acc (operationName, vars) ->
+            vars
+            |> List.groupBy id
+            |> List.map (fun (k, v) -> k, v.Length)
+            |> List.filter (fun (_, length) -> length > 1)
+            |> List.fold (fun acc (var, length) ->
```

Comment:
I am not sure if making an FSharpMap is needed, since we are only building the list for folding once. But I think using `List.countBy` is actually a better strategy:

```fsharp
List.countBy id vars
            |> List.fold (fun acc (var, length) ->
                match length, operationName with
                | 0, _ -> acc
                | _, Some operationName -> acc @ Error.AsResult(sprintf "Variable '%s' in operation '%s' is declared %i times. Variables must be unique in their operations." var.VariableName operationName length)
                | _, None -> acc @ Error.AsResult(sprintf "Variable '%s' is declared %i times in the operation. Variables must be unique in their operations." var.VariableName length)) acc) Success
```

Reply:
I also don't feel secure using a map, because we can potentially have a document with operations having duplicated names. This validation is not responsible for operation name uniqueness, so the map would only validate one of the duplicated.

Reply:
Yeah, that looks fine to me.